bindings/python/src/_pyghex/CMakeLists.txt

```diff
@@ -33,13 +33,25 @@ target_link_libraries(pyghex_obj PUBLIC ghex_common)
 # Add GPU support
 compile_as_cuda(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} SOURCES ${pyghex_source})
 
-# Add support for mpi4py if available.
+# Add MPI4PY support
 find_python_module(mpi4py)
-if (HAVE_MPI4PY)
-    target_include_directories(pyghex_obj PRIVATE "${PY_MPI4PY}/include")
-    target_compile_definitions(pyghex_obj PRIVATE -DGHEX_ENABLE_MPI4PY)
+if(NOT HAVE_MPI4PY)
+    message(FATAL_ERROR "mpi4py support is required for this build.")
 endif()
 
+# custom command to generate mpi4py.hpp via python script
+set(MPI4PY_HEADER "${CMAKE_CURRENT_BINARY_DIR}/mpi4py.hpp")
+add_custom_command(
+    OUTPUT ${MPI4PY_HEADER}
+    COMMAND ${Python3_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/generate_mpi4py_header.py ${MPI4PY_HEADER}
+    COMMENT "Generating mpi4py.hpp with mpi4py version information"
+)
+# we need custom target to establish dependency
+add_custom_target(generate_mpi4py_header DEPENDS ${MPI4PY_HEADER})
+add_dependencies(pyghex_obj generate_mpi4py_header)
+# add mpi4py include directories
+target_include_directories(pyghex_obj PRIVATE "${PY_MPI4PY}/include")
+
```

Comment:
This could probably be replaced by `configure_file` and a small extension to the `ghex_find_python_module.cmake` to retrieve the version.

```
include(FindPackageHandleStandardArgs)
function(find_python_module module)
    string(TOUPPER ${module} module_upper)

    if(NOT PY_${module_upper})
        if(ARGC GREATER 1 AND ARGV1 STREQUAL "REQUIRED")
            set(${module}_FIND_REQUIRED TRUE)
        endif()

        # A module's location is usually a directory, but for binary modules
        # it's a .so file.
        execute_process(COMMAND "${PYTHON_EXECUTABLE}" "-c"
            "import re, ${module}; print(re.compile('/__init__.py.*').sub('',${module}.__file__))"
            RESULT_VARIABLE _${module}_status
            OUTPUT_VARIABLE _${module}_location
            ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
        execute_process(COMMAND "${PYTHON_EXECUTABLE}" "-c"
            "import re, ${module}; print(re.compile('/__init__.py.*').sub('',${module}.__version__))"
            OUTPUT_VARIABLE _${module}_version
            ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
        message(STATUS "python-exe = ${PYTHON_EXECUTABLE}")
        message(STATUS "status = ${_${module}_status}")
        message(STATUS "location = ${_${module}_location}")
        message(STATUS "version = ${_${module}_version}")

        if(NOT _${module}_status)
            set(HAVE_${module_upper} ON CACHE INTERNAL "Python module available")
            set(PY_${module_upper} ${_${module}_location} CACHE STRING "Location of Python module ${module}")
            set(PY_${module_upper}_VERSION ${_${module}_version} CACHE STRING "Version of Python module ${module}")
        else()
            set(HAVE_${module_upper} OFF CACHE INTERNAL "Python module available")
        endif()
    endif(NOT PY_${module_upper})

    find_package_handle_standard_args(PY_${module} DEFAULT_MSG PY_${module_upper})
endfunction(find_python_module)
```